sky/serve/load_balancer.py

```diff
@@ -18,25 +21,135 @@
+# (see below).
+#
+# - Kubernetes Support:  The PR relies on using the `docker` command line to
+# boot an Envoy container per service.  This won't work in Kubernetes where the
+# sky controller is a pod, and thus can't run sub-containers.   There are two
+# ways this could be handled.  1) Run envoy as a process instead of a docker
+# container.  2) Switch to single-envoy per controller, and boot that Envoy in
+# the pod-spec of the Kubernetes controller.
+#
+# - Weighted Least Request: We should used the weighted least request algorithm
+# instead of round robin.  The switch shouldn't be hard, just haven't done it
+# yet.
+#
+# - Configurability:  Currently it's just configured through constants in the
+# code.  We should decide if users should be able to configure this, and if so
+# expose the relevant knobs in the service spec.
+
 
 class SkyServeLoadBalancer:
+    """SkyServeLoadBalancer: load balancer for distributing requests to Sky
+    Serve replicas.
+
+    The SkyServeLoadBalancer class serves as the base class for the the various
+    child implementations.
+    """
+
+    def __init__(self, service_name: str, controller_url: str,
+                 load_balancer_port: int) -> None:
+        """Initialize the load balancer.
+
+        Args:
+            service_name: The name of the service this load balancer serves.
+            controller_url: The URL of the controller.
+            load_balancer_port: The port where the load balancer listens to.
+        """
+
+        self._controller_url = controller_url
+        self._load_balancer_port = load_balancer_port
+        self._service_name = service_name
+
+    async def _controller_sync(self,
+                               request_aggregator: dict) -> Optional[List[str]]:
+        """ Sync with the controller once.
+
+        Contact the controller. Give it the information contained in
+        `request_aggregator`.  Receive the current set of Available replicas.
+        """
+
+        try:
+            async with aiohttp.ClientSession() as session:
+                async with session.post(
+                        self._controller_url + '/controller/load_balancer_sync',
+                        json={'request_aggregator': request_aggregator},
+                        timeout=aiohttp.ClientTimeout(5)) as response:
+
+                    response.raise_for_status()
+                    response_json = await response.json()
+                    ready_replica_urls = response_json.get(
+                        'ready_replica_urls', [])
+        except aiohttp.ClientError as e:
+            logger.error('An error occurred when syncing with '
+                         f'the controller: {e}')
+            return None
+        else:
+            logger.debug(f'Available Replica URLs: {ready_replica_urls}')
+            return ready_replica_urls
+
+
+class PythonLoadBalancer(SkyServeLoadBalancer):
     """SkyServeLoadBalancer: distribute incoming traffic with proxy.
```

Comment:
```suggestion
    """PythonLoadBalancer: distribute incoming traffic with Python proxy.
```

Reply:
I had no idea you could directly make suggestions in PRs like this.  Very cool